src/MSALWrapper/AuthFlow/DeviceCode.cs

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Authentication.MSALWrapper.AuthFlow
+{
+#if NET472
+    using Microsoft.Identity.Client.Desktop;
+#endif
+    using System;
+    using System.Collections.Generic;
+    using System.Net.Http;
+    using System.Net.Http.Headers;
+    using System.Threading.Tasks;
+    using Microsoft.Extensions.Logging;
+    using Microsoft.Identity.Client;
+
+    /// <summary>
+    /// The device code auth flow.
+    /// </summary>
+    public class DeviceCode : IAuthFlow
+    {
+        private readonly ILogger logger;
+        private readonly IEnumerable<string> scopes;
+        private readonly string preferredDomain;
+        private readonly string promptHint;
+        private readonly IList<Exception> errors;
+        private IPCAWrapper pcaWrapper;
+
+        #region Public configurable properties
+
+        /// <summary>
+        /// The silent auth timeout.
+        /// </summary>
+        private TimeSpan silentAuthTimeout = TimeSpan.FromMinutes(5);
+
+        /// <summary>
+        /// The device code flow timeout.
+        /// </summary>
+        private TimeSpan deviceCodeFlowTimeout = TimeSpan.FromMinutes(15);
+        #endregion
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DeviceCode"/> class.
+        /// </summary>
+        /// <param name="logger">The logger.</param>
+        /// <param name="clientId">The client id.</param>
+        /// <param name="tenantId">The tenant id.</param>
+        /// <param name="scopes">The scopes.</param>
+        /// <param name="osxKeyChainSuffix">The osx key chain suffix.</param>
+        /// <param name="preferredDomain">The preferred domain.</param>
+        /// <param name="pcaWrapper">Optional: IPCAWrapper to use.</param>
+        /// <param name="promptHint">The customized header text in account picker for WAM prompts.</param>
+        public DeviceCode(ILogger logger, Guid clientId, Guid tenantId, IEnumerable<string> scopes, string osxKeyChainSuffix = null, string preferredDomain = null, IPCAWrapper pcaWrapper = null, string promptHint = null)
+        {
+            this.errors = new List<Exception>();
+            this.logger = logger;
+            this.scopes = scopes;
+            this.preferredDomain = preferredDomain;
+            this.promptHint = promptHint;
+            this.pcaWrapper = pcaWrapper ?? this.BuildPCAWrapper(logger, clientId, tenantId, osxKeyChainSuffix);
+        }
+
+        /// <summary>
+        /// Gets the token for a resource.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> of <see cref="TokenResult"/>.</returns>
+        public async Task<AuthFlowResult> GetTokenAsync()
+        {
+            IAccount account = await this.pcaWrapper.TryToGetCachedAccountAsync(this.preferredDomain)
+                ?? null;
+            this.logger.LogDebug($"GetTokenNormalFlowAsync: Using account '{account.Username}'");

```

Comment:
Is this log statement still correct? Should we be referencing `GetTokenNormalFlowAsync`? 🤔

Reply:
Nope, let me update it because we don't completely try the "normal" flow.